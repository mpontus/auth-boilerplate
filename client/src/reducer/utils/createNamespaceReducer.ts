import { Action, Reducer } from "redux";

type Namespace<K extends string, V> = Partial<Record<K, V>>;

type HigherOrderReducer<S, A extends Action<any>> = (
  action: A
) => (state: S, action: A) => S;

export const createNamespaceReducer = <K extends string, V, A extends Action>(
  fn: HigherOrderReducer<Namespace<K, V>, A>
): Reducer<Namespace<K, V>, A> => (state = {}, action) =>
  fn(action)(state, action);
