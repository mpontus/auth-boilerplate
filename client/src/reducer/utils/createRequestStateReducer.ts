import { Action, Reducer } from "redux";

interface RequestState<ErrorType> {
  loading: boolean;
  success: boolean;
  error?: ErrorType;
}

const initialState = {
  loading: false,
  success: false
};

/**
 * Reusable reducer constructor for managing request state
 */
export const createRequestStateReducer = <
  ActionType extends Action<any>,
  ErrorType extends Error
>(
  reducer: (
    state: RequestState<ErrorType>,
    action: ActionType
  ) => RequestState<ErrorType>
): Reducer<RequestState<ErrorType>, ActionType> => (
  state = initialState,
  action
) => reducer(state, action);
