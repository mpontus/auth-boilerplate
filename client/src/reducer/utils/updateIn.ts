import { Action, Reducer } from "redux";

/**
 * Return a function which updates given key in an object, using the
 * provided updater function.
 */
export const updateIn = <
  K extends keyof R,
  A extends Action,
  V,
  R = Record<K, V>
>(
  key: K,
  updater: Reducer<R[K], A>
) => (state: R, action: A): R =>
  Object.assign({}, state, { [key]: updater(state[key], action) });
