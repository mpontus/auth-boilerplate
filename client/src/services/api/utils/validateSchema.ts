import * as t from "io-ts";
import { PathReporter } from "io-ts/lib/PathReporter";

export const validateSchema = <TSchema extends t.Type<any>>(
  schema: TSchema,
  data: any
): t.TypeOf<TSchema> => {
  const result = schema.decode(data);

  if (result.isLeft()) {
    throw new Error(PathReporter.report(result).join("\n"));
  }

  return result.value;
};
