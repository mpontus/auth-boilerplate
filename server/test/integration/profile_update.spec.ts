import request from 'supertest';
import { resetDb } from '../utils/resetDb';
import { initApp } from '../utils/initApp';

let expressApp: any, nestApp: any;

beforeAll(async () => {
  ({ expressApp, nestApp } = await initApp());
});

afterAll(() => nestApp.close());

beforeEach(() => resetDb());

describe('Profile update', () => {
  describe('When user is authenticated', () => {
    const seed = require('../seeds/authenticated_user');

    beforeEach(() => seed.run());

    it('should be successful', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          name: 'Victor Martinez',
        })
        .set('Authorization', `Bearer ${seed.token}`)
        .expect(202);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('When user is unauthenticated', () => {
    it('should fail', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          name: 'Victor Martinez',
        })
        .expect(401);

      expect(response.body).toMatchSnapshot();
    });
  });
});

describe('Updating email', () => {
  const seed = require('../seeds/authenticated_user');

  beforeEach(() => seed.run());

  describe('when current password is present', () => {
    it('should be successful', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          currentPassword: seed.password,
          email: 'dillondelacruz@shaw.com',
        })
        .set('Authorization', `Bearer ${seed.token}`)
        .expect(202);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when current password is invalid', () => {
    it('should fail', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          currentPassword: 'foo',
          email: 'dillondelacruz@shaw.com',
        })
        .set('Authorization', `Bearer ${seed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});

describe('Updating password', () => {
  const seed = require('../seeds/authenticated_user');

  beforeEach(() => seed.run());

  describe('when current password is present', () => {
    it('should be successful', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          currentPassword: seed.password,
          password: 'ngbA1CVl!H',
        })
        .set('Authorization', `Bearer ${seed.token}`)
        .expect(202);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when current password is invalid', () => {
    it('should fail', async () => {
      const response = await request(expressApp)
        .patch(`/auth/profile`)
        .send({
          currentPassword: 'foo',
          password: 'ngbA1CVl!H',
        })
        .set('Authorization', `Bearer ${seed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});
