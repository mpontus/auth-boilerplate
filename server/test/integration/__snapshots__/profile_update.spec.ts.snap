// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile update When user is authenticated should be successful 1`] = `Object {}`;

exports[`Profile update When user is unauthenticated should fail 1`] = `
Object {
  "error": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`Updating email when current password is invalid should fail 1`] = `
Object {
  "error": "Validation error",
  "message": Object {
    "currentPassword": "Current password is invalid",
  },
  "statusCode": 400,
}
`;

exports[`Updating email when current password is present should be successful 1`] = `Object {}`;

exports[`Updating password when current password is invalid should fail 1`] = `
Object {
  "error": "Validation error",
  "message": Object {
    "currentPassword": "Current password is invalid",
  },
  "statusCode": 400,
}
`;

exports[`Updating password when current password is present should be successful 1`] = `Object {}`;
